clang/test/Analysis/initializer.cpp

```diff
@@ -610,3 +610,44 @@ void top() {
   consume(parseMatchComponent());
 }
 } // namespace elementwise_copy_small_array_from_post_initializer_of_cctor
+
+namespace gh147686 {
+// The problem reported in https://github.com/llvm/llvm-project/issues/147686
+// is sensitive to the initializer form: using parenthesis to initialize m_ptr
+// resulted in crashes when analyzing *m_ptr = '\0'; but using braces is fine.
+
+struct A {
+  A() : m_ptr(m_buf) { *m_ptr = '\0'; } // no-crash
+  A(int overload) : m_ptr{m_buf} { *m_ptr = '\0'; }
+  A(char src) : m_ptr(m_buf) { *m_ptr = src; } // no-crash
+  A(char src, int overload) : m_ptr{m_buf} { *m_ptr = src; }
+  char m_buf[64] = {0};
+  char * m_ptr;
+};
+
+void test1() {
+  A a;
+  clang_analyzer_eval(a.m_buf[0] == 0); // expected-warning{{TRUE}}
+  // FIXME The next eval should result in TRUE.
+  clang_analyzer_eval(*a.m_ptr == 0); // expected-warning{{UNKNOWN}}
+}
+
+void test2() {
+  A a(314);
+  clang_analyzer_eval(a.m_buf[0] == 0); // expected-warning{{TRUE}}
+  clang_analyzer_eval(*a.m_ptr == 0); // expected-warning{{TRUE}}
+}
+
+void test3() {
+  A a(0);
+  clang_analyzer_eval(a.m_buf[0] == 0); // expected-warning{{TRUE}}
+  clang_analyzer_eval(*a.m_ptr == 0); // expected-warning{{TRUE}}
+}
+
+void test4() {
+  A a(0, 314);
+  clang_analyzer_eval(a.m_buf[0] == 0); // expected-warning{{TRUE}}
+  clang_analyzer_eval(*a.m_ptr == 0); // expected-warning{{TRUE}}
+}
```

Comment:
```suggestion
void test3(char arg) {
  A a(arg);
  clang_analyzer_eval(a.m_buf[0] == arg); // expected-warning{{TRUE}}
  clang_analyzer_eval(*a.m_ptr == arg); // expected-warning{{TRUE}}
}

void test4(char arg) {
  A a(arg, 314);
  clang_analyzer_eval(a.m_buf[0] == arg); // expected-warning{{TRUE}}
  clang_analyzer_eval(*a.m_ptr == arg); // expected-warning{{TRUE}}
}
```
To me it seems that these tests tried to test the branch `if (isa<nonloc::SymbolVal>(Init))` but instead of using truly symbolic value they just use ConcreteInt zero in a slightly more complex way.

(Please check that the tests actually pass with my suggested change -- although the analyzer should be smart enough to realize that a symbol is equal to itself.)

Reply:
Ok, thanks for updating the tests. 

>  I won't have time to investigate though.

No problem, the unexpected behavior of `test3Bis` is unrelated to the bug fixed in this commit (it seems to be a limitation of our modeling of the intricate details of C++ variable initialization). Just adding the FIXME is perfectly fine.